### ISSUE TITLE
How to use Groovy to construct InlineStyleTextArea and do #142's style update?

### ISSUE THREAD

**Issue by JordanMartinez** [TomasMikula/RichTextFX#143]
Hi Tomas,

This question might be better asked on StackOverflow.com because it might be outside the scope of this project. If it is, then I don't want to waste your time; I can ask it on there. On the other hand, I don't understand why this wouldn't work in its current form. So perhaps there's some underlying things going on that you would easily spot but that I can't.

I'm using Groovy and GroovyFX to design my program. Groovy's syntax doesn't understand Java 8 Lambdas; however, its Closures can replace the Lambdas and do the exact thing (though this is still something I'm figuring out).

Using the following code, I replaced 2 Lambdas with Closures (one used in InlineStyleTextArea's constructor, and one to do StyleSpans.mapStyles(mapper). However, despite using your answer in #142, updating the text with a new style doesn't change the text's color to red as it should. It remains black. Here's the code in its simplest form:

````groovy
import javafx.scene.paint.Color
import org.fxmisc.richtext.InlineStyleTextArea
import org.fxmisc.richtext.StyleSpan
import org.fxmisc.richtext.StyleSpans
import org.fxmisc.richtext.StyleSpansBuilder

import static groovyx.javafx.GroovyFX.start

// same idea as StyleInfo from RichText.java but made groovier...
class TextStyle {
    Boolean bold
    Boolean italic
    Boolean underline
    Boolean strikethrough
    Integer fontSize
    String fontFamily
    Color color

    // No constructor in Groovy means I can use a Map to construct
    //   Example:  new TextStyle(bold:true, fontSize:20...)

    String toCss() {
        StringBuilder sb = new StringBuilder()
        if (bold != null) {
            sb.append("-fx-font-weight: ${(bold) ? "bold" : "normal"};")
        }
        if (italic != null) {
            sb.append("-fx-font-style: ${(italic) ? "italic" : "normal"};")
        }
        if (underline != null) {
            sb.append("-fx-underline: ${(underline) ? "true" : "false"};")
        }
        if (strikethrough != null) {
            sb.append("-fx-strikethrough: ${(strikethrough) ? "true" : "false"};")
        }
        if (fontSize != null) {
            sb.append("-fx-font-size: ${fontSize}pt;")
        }
        if (fontFamily != null) {
            sb.append("-fx-font-family: ${fontFamily};")
        }
        if (color != null) {
            int red = color.getRed() * 255
            int green = color.getGreen() * 255
            int blue = color.getBlue() * 255
            sb.append("-fx-fill: rgb($red, $green, $blue)")
        }

        return sb.toString()
    }
}
/*
    Creates a border pane with styledArea at center.
    Clicking on button (bottom) will call replaceStyles (Closure)
 */
start {
    // Create styles
    def startingStyle = new TextStyle()  // toCss() = ""
    def replacementStyle = new TextStyle(color:Color.RED)

    // Create area and use Closure instead of Lambda.
    //   Not sure if "{startingStyle.toCss()}" creates a problem or not...
    def styledArea = new InlineStyleTextArea(startingStyle, { startingStyle.toCss() })

    def mapStyles   // declare Closure (assigned later), makes reading this easier...

    // Create Closure that simulates Tomas' style update answer
    def replaceStyles = { InlineStyleTextArea area, TextStyle oldStyle, TextStyle newStyle ->
        StyleSpans<TextStyle> originalSpans = area.getStyleSpans(0, area.getLength())

        StyleSpans<TextStyle> newSpans = mapStyles(originalSpans,

                // same basic idea but uses a Closure instead of a Lambda
                {style -> return (style.equals(oldStyle)) ? newStyle : style }
        )

        area.setStyleSpans(0, newSpans)
    }

    // Closure that copies 'mapStyles()' method exactly (except a few places for easier readability)
    //       but uses a closure instead of UnaryOperator
    mapStyles = { StyleSpans<TextStyle> originalSpans, Closure mapper ->
        int numberOfSpans = originalSpans.getSpanCount()
        StyleSpansBuilder<TextStyle> builder = new StyleSpansBuilder<>(numberOfSpans)

        StyleSpan<TextStyle> firstSpan = originalSpans.getStyleSpan(0); // there should always be at least one span
        int prevLen = firstSpan.getLength()

        // mapper.apply() normally called here; Closure used instead
        TextStyle prevStyle = mapper(firstSpan.getStyle())

        for (i in 1..<numberOfSpans) {
            StyleSpan<TextStyle> span = originalSpans.getStyleSpan(i);

            // mapper.apply() normally called here; Closure used instead
            TextStyle style = mapper(span.getStyle())

            if(Objects.equals(prevStyle, style)) {
                prevLen += span.getLength()
            } else {
                builder.add(prevStyle, prevLen)
                prevLen = span.getLength()
                prevStyle = style
            }
        }
        builder.add(prevStyle, prevLen)
        return builder.create()
    }

    stage(id: "stage", title: "Hello", primary: true, visible: true) {
        scene(width: 500, height: 500) {
            borderPane(id:"border") {
                bottom {
                    button("Replace Styles", onAction:{
                        replaceStyles(styledArea, startingStyle, replacementStyle)
                    })
                }
            }
        }
    }
    border.setCenter(styledArea)
}
````

**Comment by TomasMikula** [comment]
Hi Jordan,

I am not a Groovy programmer, but this line

```groovy
new InlineStyleTextArea(startingStyle, { startingStyle.toCss() })
```

seems to me like passing a closure over a constant `startingStyle`, ignoring the argument, while it should be something like

```groovy
new InlineStyleTextArea(startingStyle, { TextStyle style -> style.toCss() })
```

Also, I don't understand why you need to reimplement `mapStyles`.

**Comment by JordanMartinez** [comment]
Dah! That was it! Thank you, sir! It works after changing the constructor to the one you showed.

Using my previous constructor would insert the text underneath the initial style. So I assumed that what I was using was indeed working. When I tried using the previous solution you gave in updating the text, it didn't work. I thought that mapStyles might be the issue. It was only later that I thought the constructor might be written incorrectly.

Short answer: I wrote the constructor poorly, and I don't need to reimplement ````mapstyles````
